cabal-validate/main/Main.hs

```diff
@@ -0,0 +1,953 @@
+              , [TimeSummary]
+              ]
+
+      targets' =
+        concat
+          [
+            [ "Cabal"
+            , "Cabal-hooks"
+            , "cabal-testsuite"
+            , "Cabal-tests"
+            , "Cabal-QuickCheck"
+            , "Cabal-tree-diff"
+            , "Cabal-described"
+            ]
+          , optionals
+              (CliTests `elem` steps')
+              [ "cabal-install"
+              , "cabal-install-solver"
+              , "cabal-benchmarks"
+              ]
+          , optional (rawSolverBenchmarks opts) "solver-benchmarks"
+          ]
+
+      archPath' =
+        let osPath =
+              case os of
+                "darwin" -> "osx"
+                "linux" -> "linux"
+                "mingw32" -> "windows"
+                _ -> os -- TODO: Warning?
+         in arch <> "-" <> osPath
+
+      projectFile' =
+        if rawLibOnly opts
+          then "cabal.validate-libonly.project"
+          else "cabal.validate.project"
+
+      tastyArgs' =
+        "--hide-successes"
+          : case rawTastyPattern opts of
+            Just tastyPattern -> ["--pattern", tastyPattern]
+            Nothing -> []
+
+  when (rawListSteps opts) $ do
+    -- TODO: This should probably list _all_ available steps, not just the selected ones!
+    tPutStrLn "Targets:"
+    forM_ targets' $ \target -> do
+      tPutStrLn $ "  " <> target
+    tPutStrLn "Steps:"
+    forM_ steps' $ \step -> do
+      tPutStrLn $ "  " <> displayStep step
+    exitSuccess
+
+  startTime' <- getAbsoluteTime
+  jobs' <- maybe getNumCapabilities pure (rawJobs opts)
+  cwd' <- getCurrentDirectory
+  compiler' <- makeCompiler (rawCompiler opts)
+
+  pure
+    ResolvedOpts
+      { verbose = rawVerbose opts
+      , jobs = jobs'
+      , cwd = cwd'
+      , startTime = startTime'
+      , compiler = compiler'
+      , extraCompilers = rawExtraCompilers opts
+      , cabal = rawCabal opts
+      , archPath = archPath'
+      , projectFile = projectFile'
+      , hackageTests = rawHackageTests opts
+      , tastyArgs = tastyArgs'
+      , targets = targets'
+      , steps = steps'
+      }
+
+-- | Command-line options.
+data Opts = Opts
+  { rawVerbose :: Bool
+  , rawJobs :: Maybe Int
+  , rawCompiler :: FilePath
+  , rawCabal :: FilePath
+  , rawExtraCompilers :: [FilePath]
+  , rawTastyPattern :: Maybe Text
+  , rawDoctest :: Bool
+  , rawSteps :: [Step]
+  , rawListSteps :: Bool
+  , rawLibOnly :: Bool
+  , rawRunLibTests :: Bool
+  , rawRunCliTests :: Bool
+  , rawRunLibSuite :: Bool
+  , rawRunCliSuite :: Bool
+  , rawSolverBenchmarks :: Bool
+  , rawHackageTests :: HackageTests
+  }
+  deriving (Show)
+
+optsParser :: Parser Opts
+optsParser =
+  Opts
+    <$> ( flag'
+            True
+            ( short 'v'
+                <> long "verbose"
+                <> help "Always display build and test output"
+            )
+            <|> flag
+              False
+              False
```

Comment:
Does it switch between `False` and `False`?

Reply:
Yes, this preserves the behavior of `validate.sh`. I agree that this is a bit silly, but I didn't want to change anything in this PR.